affinity/cpp-20/dXXX1r0.md

```diff
@@ -73,25 +75,57 @@ The affinity problem is especially challenging for applications whose behavior c
 
 Frequently, data are initialized at the beginning of the program by the initial thread and are used by multiple threads. While some OSes automatically migrate threads or data for better affinity, migration may have high overhead. In an optimal case, the OS may automatically detect which thread access which data most frequently, or it may replicate data which are read by multiple threads, or migrate data which are modified and used by threads residing on remote locality groups. However, the OS often does a reasonable job, if the machine is not overloaded, if the application carefully uses first-touch allocation, and if the program does not change its behavior with respect to locality.
 
-Consider a code example *(Listing 1)* that uses the C\+\+17 parallel STL algorithm `for_each` to modify the entries of a `valarray` `a`.  The example applies a loop body in a lambda to each entry of the `valarray` `a`, using an execution policy that distributes work in parallel across multiple CPU cores. We might expect this to be fast, but since `valarray` containers are initialized automatically and automatically allocated on the master thread's memory, we find that it is actually quite slow even when we have more than one thread.
+Consider a code example *(Listing 1)* that uses the C\+\+17 parallel STL algorithm `for_each` to modify the entries of an `std::vector` `data`.  The example applies a loop body in a lambda to each entry of the `std::vector` `data`, using an execution policy that distributes work in parallel across multiple CPU cores. We might expect this to be fast, but since `std::vector` containers are initialized automatically and automatically allocated on the master thread's memory, we find that it is actually quite slow even when we have more than one thread.
+
+```cpp
+// NUMA executor representing N NUMA regions.
+numa_executor exec;
+
+// Storage required for vector allocated on construction local to current thread
+// of execution, (N == 0).
+std::vector<float> data(N * SIZE);
+
+// Require the NUMA executor to bind its migration of memory to the underlying
+// memory resources in a scatter pattern.s
+auto affinityExec = std::execution::require(exec,
+  bulk_execution_affinity.scatter);
+
+// Migrate the memory allocated for the vector across the NUMA regions in a
+// scatter pattern.
+std::execution::migrate(std::begin(data), std::end(data), affinityExec);
```

Comment:
I looked at this interface again and wondered: Would it ever make sense to call this for anything other than a contiguous chunk of memory?  For example, could `data` be an `std::map` here?  If not, then maybe this should just take a `span`.

Reply:
@AerialMantis `mdspan` is a bit tricky.  Even if you restrict it to 1-D, `mdspan`'s Accessor policy does not require that any raw pointer range actually exist.  Furthermore, migration happens by pages, which could be tricky for a strided `mdspan`.

Reply:
@AerialMantis Correction: `mdspan`'s Accessor policy does require that a raw pointer exist somewhere (`a.decay(p)` must return `A::element_type*`), but actually getting a range out isn't necessarily required (hence `A::offset_policy`).

Reply:
Thanks for looking into this. I think 1-D `mdspan` makes the most sense here, there would have to be a limitation that if the `mdspan` is strided then the stride length must be aligned with the page size (though that's tricky again because AFAIK that's not something available in standard C++). Do you think any further guarantees could be made on an `mdspan` by introducing a specific accessor policy for memory affinity?

Another concern that I had was that for some executors migration can be done by the OS without requiring re-allocation by the user. But in the case of discrete devices such as GPUs that's often not the case. So in this case the migrate operation would have to re-allocate and this means either providing an `mdspan` the ability to arbitrarily reallocate a container or returning a new object. But we can't return a new `mdspan` because as I understand an `mdspan` is simply a view and doesn't own the data.

I think this will require much further discussion, so for now I've made the example simply take the `std::vector` directly to avoid suggesting any particular approach. Does that sound okay?